src/cocotb/clock.py

```diff
@@ -109,52 +117,62 @@ async def custom_clock():
     """
 
     def __init__(
-        self, signal, period: Union[float, Fraction, Decimal], units: str = "step"
+        self,
+        signal,
+        period: Union[float, Fraction, Decimal],
+        units: str = "step",
+        impl: str = "auto",
     ):
         self.signal = signal
         self.period = get_sim_steps(period, units)
-        self.half_period = get_sim_steps(period / 2, units)
         self.frequency = 1 / get_time_from_sim_steps(self.period, units="us")
-        self.hdl = None
-        self.signal = signal
-        self.coro = None
-        self.mcoro = None
+        if impl == "auto":
+            impl = "gpi" if cocotb._conf.trust_inertial else "py"
+        elif impl not in ["gpi", "py"]:
+            raise ValueError(f"Invalid clock impl '{impl}'")
```

Comment:
```suggestion
            raise ValueError(f"Invalid clock impl {impl!r}")
```
We might also print the supported impls here.